### PR TITLE
[WIP] Auto-tag components

### LINKED ISSUE #17: data attribute bindings on Ember view helpers
Great work on this add-on; elegant solution to the problem. We have been using namespaced class names but it feels really icky to see those classes in our production code.
##### This is not really a bug, but more of a gotcha if you are not paying attention or don't know how Ember view helpers work...

According to the [2.4.x Docs](https://guides.emberjs.com/v2.4.0/templates/binding-element-attributes/#toc_adding-data-attributes) data attributes are not supprted by default. In order to make this addon work with them you have to manually reopen the classes and specifically list any data attributes you are going to add. This becomes obvious pretty much immediately since one of the first things we typically want to do in an acceptance test is target form elements and link-tos.

While I understand reopening classes is not a big deal, it would be nice if there as a way for the addon to override this behavior for all the built in Ember view helpers when it is installed. Not sure if that is possible but I would be happy to help brainstorm and implement a way to accomplish it. Maybe in the config users could pass a hash of the helper and data attributes... If I have time this weekend maybe I'll try to implement something.

### PR DESCRIPTION
**Pull request by marcoow**
This automatically adds `data-test-component="<component-name>"` attributes to all components rendered to allow to easily select any component instances with `testSelector('component', 'component-name')`.

Closes #17 

## TODO

- [ ] remove initializer from the production build
- [ ] strip all `data-test-` attributes (both ones defined in the component's JS as well as the ones assigned in templates)
- [ ] add tests

### CODE REVIEW

File: addon/initializers/add-test-selectors-to-components.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  Component.reopen({
+    init() {
+      let attributeBinding = `-data-test-component-selector:data-test-component`;
+      this.attributeBindings = (this.attributeBindings || []).concat([attributeBinding]);
+
+      return this._super(...arguments);
+    },
+
+    '-data-test-component-selector': computed(function() {
+      return nameOfComponent(this, config);
```
Comment (bcardarella) [inline addon/initializers/add-test-selectors-to-components.js]: perhaps I'm just not following, but I don't see anywhere to add custom selectors. This seems to only add a selector based upon the component name.

**Comment by bcardarella** [comment]
re: ` make sure this works with pods` according to Core pods were never an officially recommended app structure and will be going away pending RFC 143. You may want to rethink supporting them.

**Comment by marcoow** [comment]
That's true about pods but there are lots of projects out there using them so if we can support them I think we should. 

This will likely have to be adopted for the module unification changes as well.

File: addon/initializers/add-test-selectors-to-components.js (hunk 1 above)
Comment (marcoow) [inline addon/initializers/add-test-selectors-to-components.js]: Yes, for now I want to support attributes with the component name and then for all `data-test-*` attributes afterwards.

**Comment by lukemelia** [comment]
FWIW, if this lands I would avoid relying on it in my tests. My reasoning is that test selectors should be independent from other implementation concerns to allow for refactoring without test breakage.

In addition to changing CSS classes, element types, and HTML structure, the list of things you should be able to do without causing acceptance tests to fail should include changing the names of your components.

**Comment by bcardarella** [comment]
I agree with @lukemelia's point here, I don't understand the desire to default to component names. It doesn't seem correct and in many cases would be non-obvious on what the component proper name was. Having an assignable unique ID is a must have IMO.

**Comment by bcardarella** [comment]
FWIW, we've actually deviated from my original snippet and have tightened the API up a bit:


```js
import Ember from 'ember';
import config from '../config/environment';

const {
  get,
  getWithDefault,
  Component
} = Ember;

if (config.environment !== 'production') {
  Component.reopen({
    init() {
      let attrs = Object.keys(this.attrs || {});
      let dataTestAttrs = attrs.filter((attr) => {
        return attr.indexOf('data-test-') === 0;
      });

      let skipTestSelector = get(this, 'skipTestSelector');

      if (dataTestAttrs.length > 0 && !skipTestSelector) {
        let attributeBindings = getWithDefault(this, 'attributeBindings', []);
        this.attributeBindings = attributeBindings.concat(dataTestAttrs);
      }

      return this._super(...arguments);
    }
  });
}
```

So it accepts the following:

```hbs
{{foo-bar data-test-foo-bar="baz"}}
```
results in
```html
<div data-test-foo-bar-baz="bar"></div>
```
and
```hbs
{{foo-bar data-test-foo-bar=true}}
```
results in
```html
<div data-test-foo-bar></div>
```

We've found this API to be concise and very flexible.

**Comment by bcardarella** [comment]
Also, I believe having the option of `skipSelector` is necessary. For example, you may pass into a parent component the test selector attribute, but only want it applied to child components. In this event the skip selector would be applied to the component prototype.

**Comment by marcoow** [comment]
> In addition to changing CSS classes, element types, and HTML structure, the list of things you should be able to do without causing acceptance tests to fail should include changing the names of your components.

Yeah, that's actually a good point. I was originally thinking mainly about test cases like _"when the server responds with x things, x things should be rendered"_ which might be easily implemented when each component had a `data-test-` selector out of the box that would default to the component name. It is correct of course that this would encourage tests that depend on internals as the component name.

I'm changing this so that we'll just bind all `data-test-` attributes as @bcardarella suggests. There doesn't seem to be a good way to have `data-test-` attributes added to components by default.

> Also, I believe having the option of skipSelector is necessary. For example, you may pass into a parent component the test selector attribute, but only want it applied to child components. In this event the skip selector would be applied to the component prototype.

I'm not sure I understand the use case of this. Can you explain?

**Comment by bcardarella** [comment]
> I'm not sure I understand the use case of this. Can you explain?

Yup! So imagine a scenario where you have a form-for component:

```hbs
{{form-for user as |f|}}
  {{f.input "name"}}
{{/form-for}}
```

The `{{f.input "name"}}` line could render a wrapper `div` that contains the `input` element:

```html
<div class="input">
  <input id="ember123" type="text">
</div>
```

From the API available for that form-for component the only way to pass in the `data-test-` value is:

```hbs
{{f.input "name" data-test-input="name"}}
```

Which would render:

```html
<div class="input" data-test-input="name">
  <input id="ember123" type="text">
</div>
```

Which doesn't allow us to attach to the correct element. What we have done in our app is re-open the `f.input` component template and pass the value of `data-test-input` into the corresponding input:

```hbs
{{input value=property data-test-input=data-test-input}}
```

so now without the `skipSelector` we get:

```html
<div class="input" data-test-input="name">
  <input id="ember123" type="text" data-test-input="name">
</div>
```

Which of course gives us too many elements when we want to select them. So re-opening the `f.input` component and adding:

```js
skipSelector: true
```

will render:

```html
<div class="input">
  <input id="ember123" type="text" data-test-input="name">
</div>
```

Which is what we want.

Now, while I'm writing all this I am realizing that we could have just done:

```js
let selector = `${testSelector("input", "name")} input`;
```

Thoughts?

**Comment by marcoow** [comment]
Yeah, you could actually just scope selectors which I think result in quite a bit simpler code overall. You could of course also scope 2 or more `data-test-` selectors if you want to avoid selecting the `input` by its tag name:

```js
let selector = `${testSelector("input", "name")} ${testSelector("input")}`;
```

**Comment by bcardarella** [comment]
@marcoow confirm, that's one of those times where I literally just rubber ducked myself into the better solution on the GH issue. You can ignore the `skipSelector` stuff :)

**Comment by bcardarella** [comment]
@marcoow I have an integration test for the functionality extracted from our app:

```js
import Ember from 'ember';
import { moduleForComponent, test } from 'ember-qunit';
import hbs from 'htmlbars-inline-precompile';
import testSelector from 'web-experience-app/tests/helpers/ember-test-selectors';

const {
  Component
} = Ember;

const TestComponent = Component.extend();

moduleForComponent('test-component', 'Integration | Component | class extension', {
  integration: true,
  beforeEach() {
    this.register('component:test-component', TestComponent);
  }
});

test('it will render `data-test-` attrs as props when `true` is passed', function(assert) {
  this.render(hbs`{{test-component data-test-foo=true}}`);
  let selector = testSelector('foo');
  let element = this.$(selector);

  assert.equal(element.length, 1);
});

test('it will render `data-test-` attrs with associated value when passed', function(assert) {
  this.render(hbs`{{test-component data-test-foo="bar"}}`);
  let selector = testSelector('foo', 'bar');
  let element = this.$(selector);

  assert.equal(element.length, 1);
});
```

**Pull request closed without merging**